cpp/serve/model.h

```diff
@@ -20,15 +22,110 @@ using tvm::Device;
 using namespace tvm::runtime;
 
 /*!
- * \brief Create the runtime module for LLM functions.
- * \param reload_lib The model library. It might be a path to the binary
- * file or an executable module that is pre-loaded.
- * \param model_path The path to the model weight parameters.
- * \param device The device to run the model on.
- * \return The created runtime module.
+ * \brief The model module for LLM functions.
+ * It runs an LLM, and has an internal KV cache that maintains
+ * the history KV values of all processed tokens.
+ *
+ * It contains the following functions:
+ *
+ * Model related:
+ * - "token_embed": take token ids as input and return the embeddings,
+ * - "batch_prefill": take embedding of a single sequence
+ * as input, forward the embedding through LLM and return the logits,
+ * - "decode": take the embeddings of the last-committed token of an
+ * entire batch as input, forward through LLM and return the logits
+ * for all sequences in the batch,
+ * - "softmax_with_temperature": take logits and temperatures, return
+ * probabilities.
+ *
+ * KV cache related:
+ * - "create_kv_cache": create the KV cache for this module,
+ * - "add_new_sequence": add (declare) a new sequence in the KV cache,
+ * - "remove_sequence": remove a sequence from KV cache.
+ *
+ * ... and some other auxiliary functions.
  */
-MLC_LLM_DLL tvm::runtime::Module CreateModelModule(TVMArgValue reload_lib, String model_path,
-                                                   DLDevice device);
+class Model : public Object {
+ public:
+  /*!
+   * \brief Create the runtime module for LLM functions.
+   * \param reload_lib The model library. It might be a path to the binary
+   * file or an executable module that is pre-loaded.
+   * \param model_path The path to the model weight parameters.
+   * \param device The device to run the model on.
+   * \return The created runtime module.
+   */
+  TVM_DLL static ObjectPtr<Model> Create(TVMArgValue reload_lib, String model_path,
```

Comment:
No need to use `ObjectPtr<Model>`. Instead you may forward-declare `class Model` and then use it

Reply:
Gonna switch to having both Object and ObjectRef in the incoming push.